RFC: Crate changelogs

newpavlov opened the pull request:
[Rendered](https://github.com/newpavlov/rfcs/blob/master/text/0000-crate-changelogs.md)

This RFC proposes addition of changelog support to `cargo` and crates.io.

The main proposal is to add `changelog` field to `[package]` section of `Cargo.toml` which can point either to changelog file supplied with the crate, or to external resources. (e.g. GitHub releases or custom websites)

In the first case, changelog file should follow the minimal set of rules. Also this RFC specifies additional check for `cargo publish`, so crate authors will not forget to update changelog before publishing the new version.

Changelog rendering on the crates.io can be done through additional link on top bar:
![1](https://user-images.githubusercontent.com/329626/29746037-86920e8a-8ad5-11e7-828c-4d32f6ac4cf2.png)

This proposal initially also included a multi-file approach to handling changelogs. (most of the initial comments discuss this part) While I believe it would be good for ecosystem to define stricter conventions around changelogs with enhanced machine readability, but based on response, I've moved this part to a separate (closed) [RFC](https://github.com/rust-lang/rfcs/pull/2131).

Lokathor commented:
I'm not professional enough to write changelogs. Will cargo and/or crates.io be able to automatically generate an API diff from one version to the next and spit out that as the change log?

newpavlov commented:
@Lokathor 
Certainly not as part of this RFC. But this functionality can be introduced by subsequent RFCs in the form of additional `cargo changelog` subcommands.

nvzqz commented:
Glad my post could help get this discussion going! :)

I don't think that having a changelog be a commit diff is a good idea. There's lots of irrelevant messages from commits like with merge commits.

@lokathor If you don't believe you're professional enough to write a changelog, open an issue to ask for help with it. The issue will bring attention to the lack of changelog to people who see the repo.

nvzqz commented:
How should a changelog address a project that has been deprecated in favor of another project?

And if not in favor of another, it should address that the author may not have enough time to support it. Or maybe the project had unsafe effects (accidental malware?) that people want to study rather than used in legitimate software?

boustrophedon commented:
@nvzqz Wouldn't that sort of stuff just go in the README? Changelogs (as promoted by keepachangelog) don't usurp readmes.

nvzqz commented:
@boustrophedon the changelog described has an Unreleased section. It would just be a message stating that no more releases should be expected.

tblair commented:
I'm uncomfortable with Rust tooling taking an opinionated view towards changelogs, especially when that opinion marginalizes a project like [clog](https://crates.io/crates/clog) that's been part of the Rust ecosystem for a long time now. I can definitely see why people would prefer the keepachangelog.com approach, but I don't think Cargo should make that decision for project authors. This would be especially galling to any project that has put the time and effort into maintaining a conventional commit history.

This is an area where there are multiple specifications and whatever Rust tooling gets built should allow project authors to choose whichever approach they prefer.

For anyone unfamiliar, [here](https://conventionalcommits.org/) is the spec for the conventional commit method of handling changelogs and [here](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit) are the Angular commit guidelines on which its based.

sunjay commented:
I think changelog support would be awesome, but I don't get the need for the custom split-file format. As stated in the RFC, this deviates from the norm. I don't know if anyone would think to look for a changelog folder. I certainly wouldn't want to have one in my repo because I prefer the single file changelog. 

I like everything about adding the changelog link to crates.io and I'm in favor of having the key for that in Cargo.toml, but I'm not sold on the split changelog thing. I don't see why the changelog needs be generated at all. 

I know maintaining a changelog by updating it's links and adding new sections can be a pain, so maybe an external tool that makes that awesome would be a good idea. Just working on changelogs doesn't seem like something that should be part of cargo. 

Thanks for making the RFC! I'm glad someone is talking about this. :)

vorner commented:
I like the intent to add a changelog to crates.io and to `Cargo.toml`. I even agree with requiring certain format (it makes sense, if it needs to be rendered somewhere).

However, I very much dislike the split into files. I keep changelogs with my crates (at least once the API settles a bit), but I prefer having all the changes in one file. It's easier to search for something. And when I download a repository (or open it at github), having to either generate the changelog locally or searching multiple files feels like a lot of added work for no real gain (and keeping a generated file in git is also wrong). Also, having to be online for generating the changelog feels somewhat inconsistent with the fact that almost everything cargo does works offline (except when adding and downloading dependencies).

What work is actually saved by that split? Adding the dates? Moving the lines from unreleased section to the current version? If the changelog has a standardized format (MD with the given sections), any tooling should be able to handle it in a single file as well.

mehcode commented:
:+1: on prescribing a _general_ format as loose as possible
 - Headers with version/date
 - Bullet lists with updates
 - Bullet lists can be optionally categorized 

I'd like to define the layout enough that we could do [this](https://gemnasium.com/gems/activerecord-jdbcsqlite3-adapter) on crates.io but no more.

I strongly dislike the tool that generates a changelog from a folder structure being officially part of Rust. I'd rather see that tool compete with clog as separate crates.

est31 commented:
I'm not really a fan of adding yet another folder. One file should be enough IMO. If I look at [the changelog](https://github.com/RustAudio/lewton/blob/master/CHANGELOG.md) I keep for my `lewton` crate, I see some releases with only two bullet points. Making a file for only one or two bullet points is a waste IMO.

Generally I think we should be as liberal as possible with the changelog format, and not dictate one.

bluss commented:
I agree with @est31. From the author's perspective, a file per release is neat and simple. From the user's perspective, when I look at a change log, I am probably reading it like a timeline: I'm going to read everything from some point and up until the current state, so I'll read through multiple releases, and a single file is then much easier to read.

Of course for big enough projects, splitting the change log per release again becomes reasonable.

sunng87 commented:
+1 for keeping changelogs in a single file. Also I think a changelog needs to be written by hand to include information curated and prioritized.

In recent version of [handlebars](https://github.com/sunng87/handlebars-rust), I put unreleased changes in an *Unreleased* section, when the change is merged into master. And I configured [cargo-release](https://github.com/sunng87/cargo-release) to replace *Unreleased* with new version number. Then the changelog is ready.

I knew a lot effort has been put into changelog auto generation. But this is my best workflow by far.

_Sent from my Google Nexus 9 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

flosse commented:
> +1 for keeping changelogs in a single file. Also I think a changelog needs to be written by hand to include information curated and prioritized.

+1

arthurprs commented:
I have mixed feelings about this. What about just inserting a changelog link on crates.io that shows the rendered `CHANGELOG{.md}` if present?

ruuda commented:
It looks like the consensus so far is that a single file is preferred over multiple files. What the format of that file should be is less clear.

 * It could be strictly specified, which can be useful for machine readability, but which would force a format onto projects that might not be suitable for them.
 * Projects could be free to choose the format which makes sense for them, just like they can with a readme.

I would prefer the second option. What a good changelog looks like depends a lot on the size, release schedule, and purpose of the project.

mathstuf commented:
I'd love to have the changelog rendered, but I'd rather it just be "we will render any of these paths in crates.io" and how it gets there is uncared about (additional tooling is fine). I really don't like the `type: ` prefix specs because I feel it wastes precious subject space (I prefer `area: ` prefixes because they are easier to sort/skim on).

The way I usually do release notes is that each PR which needs one drops a new file in a specified directory and at release time, they get collated into a single file and deleted. This stops conflicts from happening, supports cherry-picking into backport branches easily with no change in process, and allows for the maintainer to collate and group release notes as necessary or wanted.

newpavlov commented:
Thank you for your responses!

While it was not my intention I think emphasis on a new multi-file approach was too strong. I've moved it into a separate RFC and updated this one to a minimal version.

One thing which I think is very important is to have at least minimal degree of machine readability and safety check in `cargo publish` based on that. So `cargo` will be able to help library authors to remember to update their changelogs.

mathstuf commented on text/0000-crate-changelogs.md:
```diff
@@ -0,0 +1,123 @@
+[alternatives]: #alternatives
+
+This proposal defines minimal format for changelog, leaving as much flexibilty
+for crate authors as possible withoout droping the basic machine readability.
```
typo: "dropping"

jimmycuadra commented:
The `changelog` field in the Cargo manifest should just be a URL like the `documentation` field. It doesn't make sense to impose any particular mechanism for change logging. As proposed in the RFC, this would not be useful to me because I use GitHub's releases page instead of a file in the repository. I've also seen some projects use detailed commit messages on Git tags instead of a file. I don't think users really care what format the change log is in, just that there is one and that they can find it.

newpavlov commented:
@jimmycuadra 
I don't agree with your position, I think it's quite important to have a coherent way of showing, instead of using links to a dozen of external resources. Also projects which would like to use a different approach always can specify it in the readme.

But either way I will mention your proposal as an alternative solution.

killercup commented:
My comment on your other changelog RFC is also valid for this one :) https://github.com/rust-lang/rfcs/pull/2131#issuecomment-325199451

bb010g commented:
I don't think external links are a good idea. Changelogs in the repository are useful both for their offline access and for the interface options they allow. Crates.io could gain nicer rendering with a standard format in the future with files, but that requires using GitHub/GitLab/wherever's API and constantly tracking changes for each platform we support for external links.

withoutboats commented on text/0000-crate-changelogs.md:
````diff
@@ -0,0 +1,129 @@
+changelog = "CHNAGELOG.md"
+```
+
+There is several requirements for changelogs:
````
What do you mean "requirements"? What happens if I write a changelog that doesn't conform to this?

newpavlov replied:
It will result in the error when you'll try to `cargo publish`, crates.io should deny upload of such crates. I will update RFC to state this explicitly.

nrc commented:
SGTM! For another point of reference, the VSCode marketplace takes a similar approach, but always looks for `CHANGELOG.md`, i.e., you can't specify a filename. It also doesn't do any fancy processing of the file, it just renders the markdown. That seems fine to me, having formatting requirements seems to make things more complex without bringing a lot of benefit to users.

withoutboats commented:
I would think we should do the same thing we do for readmes: if there's a file called CHANGELOG.md, render it (but in this case, on a secondary page calle changelog). That makes this an entirely crates.io feature and any existing crates with a CHANGELOG.md will automatically start displaying a changelog on crates.io.

I don't understand the portions of this that are opinionated about formatting.

withoutboats commented:
Actually, READMEs do require a readme key. Let's just do that, then; a key for your changelog file and we'll render it using the same rules for READMEs.

newpavlov commented:
@withoutboats 
I believe changelog should be structured and it's correctness should be checked by cargo/crates.io. In future we could want to do fancier things compared to a simple rendering, e.g. filtering out older or pre-release versions, showing only minor/major version changelogs. AFAIK proposed rules are fully compatible with most common formats used in the wild. (contr-examples are welcome!)

pornel commented:
I have a [use-case for machine-readable changelogs](https://github.com/rust-lang/cargo/issues/1369) — display a list of breaking changes when a Cargo dependency is upgraded.

sunjay commented:
> I have a use-case for machine-readable changelogs — display a list of breaking changes when a Cargo dependency is upgraded.

This sounds pretty cool, but I wonder, is there a situation where a short message would be used where the entire changelog wouldn't suffice? I get that it's a little more convenient maybe to see just a single version or something, but if we render the entire changelog, we can see all the changes in all the versions as we go. No need for any special files or commands.

In the comment you linked to, you described a use case where someone jumps from 1.0 to a 2.0 release and with each release from then on they see some sort of small message summarizing the changes. That sounds like it would work to display the contents of the changelog for that specific version. However, as it's been pointed out, it would also work and be easier and be more customizable to display the whole thing. The effort being saved to scroll a bit to find what you're looking for or to have to click on a link or open a file doesn't seem to providing very much benefit. 

Maybe I just don't see it yet?

pornel commented:
Importance of changelog entries varies *a lot*, so I think filtering is critical. 

For patch releases the changelog is going to be only non-actionable trivialities. This doesn't have to be displayed unless someone asks for it (if they're just fans of the project and curious how it's getting along).

But once in a while the changelog is going to contain list of breaking changes for a major version, and user will need to pay attention. Do they use the broken API? Do they need to migrate their code/config/data files?

I think displaying actionable information for major breaking changes will be helpful. Users will see exactly what they need to see, at the right time, right when they update the project.

Compare:

```sh
cargo update
```
> Upgraded db to 2.0: switched to a new config file format

vs

```sh
cargo update
```
> Upgraded db to 2.0: 
> * improved speed of foobar routine
> * switched to a new config file format
> * added Hungarian translation of user manual
> * fixed compatibility with GNU/Hurd

---- 

In the first case you see something happened, and you can decided whether to do something about it. In the second case more text turns it into "yada, yada, yada, tl;dr".

mehcode commented:
@sunjay @pornel 

Even more compelling is we could have `cargo update --interactive` which could let us see those changes easily before we say "yes, update please".

pornel commented:
BTW, Debian is quite serious about their changelogs, and they have years of practice and tooling around them. Some interesting bits:

* Urgency: https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Urgency

   Updates, regardless of semver, can be marked as as critical. There are cases when an urgent security fix is also technically a breaking change (e.g. disables vulnerable feature). In such situation my Debian servers automatically e-mail me the changelog entry! (but never spam me with low urgency changelogs).

* Changelogs can grow big: https://www.debian.org/doc/debian-policy/ch-docs.html#s-changelogs

   > these files should be installed compressed using gzip -9, as they will become large with time even if they start out small.

pornel commented:
Also look at Node's changelog: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V8.md#8.3.0

newpavlov commented:
@pornel 
I would be happy to add additional requirements and structure to changelog, but as you can see even the minimal set of rules have been met with some push-back.

Regarding your use case, I think this RFC enables it, as `cargo` will be able to filter out only major version changelogs (or minor for `0.*.*`). As for default rendering on crates.io, should we render everything or only important parts, this RFC does not specify it, so I guess we'll start with plain rendering without any filtering, which we will be able to change in future.

One idea to enhance changelogs would be to add `changelog-format` with default `simple` value which will correspond to format described in this RFC, possible examples would be:
- `simple:gzip` for compressed simple format
- `keepchangelog` for format specified by the website (fixed headline style and grouped changes)

So it will allow to use stricter formats, but it will introduce additional complexity to `cargo` and I don't believe it's a good solution.

Other solution would be to define hooks to installable `cargo` changelog tools. (with default tool which will parse `CHANGELOG.md` files) But I think it will require to split changelog from the rest of crate on crates.io, as tool can work without explicit changelog file. Also it rises many question about how crates.io will work with different formats produced by such tools. Obvious solution would be to define single (machine readable) format to which those tools will have to convert changelogs, but I am not sure if such approach will be welcomed by community.

tapeinosyne commented:
I reckon introducing a `changelog` key – be it a file, a path, or an url – to the Cargo manifest format is all that should be done, at least for now. Perhaps Cargo could *lint* for missing `changelog` keys, especially if the crate version is above `1.0`. That I would fully support.

Beyond that, it feels premature to mandate or speculate on rules – let alone machine-readable formats – at a time where changelogs are yet to become common practice amongst Rust projects; current conventions are too few and too loose to make them into obligations for library developers.

newpavlov commented:
@tapeinosyne 
Could you please provide examples of changelog formats used in the wild by crates and which do not fall under rules described in the RFC? (well, except of github releases, which are kind of changelog)

Problem with not specifying any basic requirements for changelog format is that when we'll want to enhance tooling around changelogs, we'll have to break backward compatibility (be it in cargo or/and crates.io), which is always not a fun thing.

>at a time where changelogs are yet to become common practice amongst Rust projects

As I see it, this argument works better in a backward direction. If it's not a common practice yet then we can establish a good standard, which people will be able to follow, because they'll start to keep changelog after introduction of this feature.

withoutboats commented:
We don't have to break backwards compatibility to provide more enhance dfeatures. We just only provide those features *if* you conform to whatever format we require to enable those features. e.g. if your section headers are version numbers, we will hide the yanked ones; if they're not, we don't.

newpavlov commented:
@withoutboats 
But it means we'll have a significant incoherence in the ecosystem, some will support this feature, others will not. It will make working with changelogs harder, as tools will have to know about different features and their combinations. So in result I think it will make overall experience of using changelogs significantly worse.

sunjay commented:
I don't think the changes would be so significant. I don't expect most people to make wildly different changelogs. Most changelogs I've seen seem to be fairly regular. Many people adopt the [keepachangelog](http://keepachangelog.com/en/1.0.0/) format. I imagine that it would work something like this:

1. We add a changelog key, people start to see and like having changelogs on crate pages (just like with READMEs)
2. People copy the changelog formats they like and use them on their projects
3. After a few months or years (whenever the need arises for more features), we look at the most popular formats and start to add lints
    * At this point we have a lot of information so we can decide on a format that fits most people and makes upgrading not terribly hard
4. People upgrade whenever they want to the new format

I agree with @tapeinosyne that having a strict format right now seems very premature. It makes a lot of sense to see what people start to use and then base the format off of that. Basic support is really important, but I still don't see the need for the standard format.

newpavlov commented:
@sunjay 
Why wait anything if we can go and see crates which keep changelog right now? From what I've seen they fall under minimal rules (don't know why you call them "strict", I wouldn't even call it a "standard format") described in this RFC. At least no one provided contr-examples yet. And people who don't yet keep changelogs will have no problems to follow them.

I believe it's better to make it more or less right with the first try, then later try to deal with consequences of overly relaxed rules.

BurntSushi commented:
> At least no one provided contr-examples yet.

I haven't been following this thread closely, but does this follow the format? https://github.com/BurntSushi/ripgrep/blob/master/CHANGELOG.md

newpavlov commented:
@BurntSushi 
Yes, it does. It's just that instead of `#`  you use underline style to define headlines. (probably should mention in the RFC that it's fine to use both methods)

mehcode commented:
I'd like to see at least the following minimally specified:
 - Consistent level markdown headings must start with a semver. To parse, check the first heading of each level for valid leading semver. When found, assume that level is our "version level".

```markdown
# Anything goes here

Anything goes here.

## 1.2.3

Anything goes here.

## 0.2.3

Anything goes here.

## 0.0.3

Anything goes here.
```

_or_


```markdown
# 1.2.3

Anything goes here.

# 0.2.3

Anything goes here.

# 0.0.3

Anything goes here.
```

_or_

```markdown
### 1.2.3

Anything goes here.

#### Bug Fixes

More anything goes here.

### 0.2.3

Anything goes here.

### 0.0.3

Anything goes here.
```

---

I don't think we should try and specify the format inside each chunk. I just want crates.io to be able to extract a version chunk and show it.

newpavlov commented:
@mehcode 
I've tried to define exactly that you've demonstrated in your examples. Probably wording was not clear enough. Later I will update RFC with your examples.

withoutboats commented:
> But it means we'll have a significant incoherence in the ecosystem, some will support this feature, others will not. It will make working with changelogs harder, as tools will have to know about different features and their combinations.

We've already established that breaking changes are a no go, which means the guarantees in this RFC are all the guarantees we could enforce. The guarantees in this RFC are that headers begin with a version number. Can you explain how having crates on crates.io which don't follow this format will make it hard for tools to process change logs?

At minimum they can just report an error on those crates - unfortunate but not a big loss. At maximum they can accept slightly more inputs. For example, a more liberal tool might accept headers that *containing* a semver version in their header, possibly not in the initial position, to handle people who might write "Release 1.2.0" or "<Release Name> (2.0.0)" in their headers.

Moreover, this RFC defines the changelong syntax in terms of markdown. While today crates.io only renders markdown, there are feature requests to render other formats types, such as asciidoc. I feel quite strongly that, at least as far as crates.io is concerned, any features should be based on the header tags in the rendered HTML output, not in the syntax of the source text.

carols10cents commented:
I agree with the goals of this RFC, in that making changelogs more visible will encourage more people to keep them!

I'd like to suggest, however, that the `changelog` key in `Cargo.toml` could go to a URL or to a file in the `.crate` package. In both cases, if this key was present, we would show a link on a crate's page to changelog. If the `changelog` key specifies a file, crates.io would render that using the same mechanisms/support as we recently added for readmes, and the link would go to that file. If the key was a URL, it would be a normal external link.

This would support the case that @jimmycuadra [raised](https://github.com/rust-lang/rfcs/pull/2129#issuecomment-325198349) of using Github's releases page as the changelog, and would also support people who have a whole website for their crate of which changelogs is one part.

I would be in favor of this proposal if that change was made, but currently I'm skeptical of the (lack of) flexibility the current proposal has.

tapeinosyne commented:
> Could you please provide examples of changelog formats used in the wild by crates and which do not fall under rules described in the RFC?

Excepting the prohibition of `Unreleased` or similar, I expect that most projects would fall under the proposed rules. However, there aren't very many in the first place, and "most" of "not many" is not enough, in my eyes, to warrant standardization as a leading step.  Standards are expensive demands, and direct experience is valuable. We can draw ideas from other communities, but we need more insight into our own before doing anything so bold as demanding.

> As I see it, this argument works better in a backward direction. If it's not a common practice yet then we can establish a good standard, which people will be able to follow, because they'll start to keep changelog after introduction of this feature.

I agree that newly recommended practices should come with *guidelines*, but not a standard; not for something like changelogs, where needs and wants can vary considerably across projects, as others already remarked. 

I also have a scruple: changelogs, like readmes, are traditionally authored in plaintext. From plaintext I expect freedom of form. For a tool to mandate plaintext but *demand* adherence to external conventions, however reasonable they may be, would be a small treachery: the failure to meet long-established expectations. (Markdown is valid plaintext, at least for the purpose of this comment.) Perhaps the tool could *offer* conventions which guarantee additional capabilities – much like GitHub may render commit hashes as links in release notes without demanding anything much of maintainers. Again, however, I am not persuaded we should lead with that.

newpavlov commented:
I've updated RFC and added ability link to changelogs published on external resources. Hope this will address most of the concerns regarding perceived strictness of the rules, as it will provide additional flexibility if desired. But I believe that rules for explicitly provided changelog files and basic `cargo publish` support should be left as described.

vorner commented:
I think the current set of rules makes some kind of sense. I mean, changelog is technically free form and every author can do whatever they like, but:

* It can hardly be called a changelog if it is not separated by versions.
* The rules need to be followed *only* if you want your changelog to be rendered on crates.io. If you want to link to it (then you provide an url), have it as a .bmp image in your repository (then you don't provide the changelog key), etc, you're free to do that, only crates.io won't provide the service of rendering and hosting it. And if you ask it to render it for you, it is sane to provide a format it understands.
* It is always possible to relax the rules in the future.

There's only the question of the unreleased section. On one hand, it makes sense, because at the time you publish the crate, nothing is unreleased, so that's a lint you forgot to move the entries to specific version. On the other hand it might make sense to keep it there in the meantime, not delete and recreate it every time a release is made. There's the possibility of allowing only empty unreleased section, but that feels a bit adhoc.

Also, I was wondering, could `cargo new` crate an empty changelog, as a hint that it's a good idea to have one?

newpavlov commented:
>On the other hand it might make sense to keep it there in the meantime, not delete and recreate it every time a release is made.

Good point, I will add it to unresolved questions. Personally I do not like that much idea to keep it, as I think initially changelog will be rendered in the same way as readme, so it will be strange to see "unreleased" section in it. Or we'll need to add filtering which will feel even more adhoc.

>Also, I was wondering, could `cargo new` crate an empty changelog, as a hint that it's a good idea to have one?

I think it сould be done as a separate RFC. Because for example currently `cargo new` does not create empty readme file, while arguably good readme is more important than changelog, especially for potential crate users.

bluss commented:
Some existing crates have a changelog in a different format (restructuredtext) so anything that allows such things are a plus to me. It's not required that it can render it, but it requires that non-markdown isn't forced through a markdown renderer.

newpavlov commented:
@bluss 
As I wrote in the RFC it's definitely possible to extent number of supported markup languages. But to simplify the process I think it should be done through issues for `cargo` and crates.io after implementation of this RFC using MarkDown. If we'll support validation it will be logical to synchronously add rendering support to crates.io as well.

dbrgn commented:
As the author of the first linked blogpost I very much approve of this RFC in general :) I have a few remarks though.

In my opinion some things in this RFC are too strict:

> As a consequence of the last requirement changelog must not contain "Unreleased" section

I think that's overly strict and that Cargo should not add such a requirement. It will only result in workarounds, like not adding a heading for the unreleased changes.

> All crate versions listed in the changelog must have been published earlier on crates.io, except the first entry

Why is that? It only makes it harder for projects that have had a few releases before their first crates.io release to publish. I see no advantage in adding this restriction.

In general, I'd either go for a specification that makes the format machine readable (to make it easier for automated tooling), or keep the format unspecified (except for markdown, so it can be rendered). I tend to favor the second approach. It can also be a gradual approach, where advanced rendering is done *if* the format corresponds to certain rules. As others in this thread have written, we could wait until some conventions form and until some tooling exists. If - for example - cargo would start to show changes on package updates, people would probably automatically start to format their changelogs in a way that they're supported by the tooling.

Oh, and I like the fact that you can link an external changelog, as an escape hatch for people that don't want to conform to the rules required for an explicit changelog file.

Regarding one of the unresolved questions:

> Should we check correctness and/or availability of URL in the changelog field?

I would definitely do a check and warn the user if the URL isn't reachable (e.g. if a status code different than 2xx or 3xx is returned). But I'd provide a way to publish even if that URL is (temporarily?) unavailable.

(PS: I found two typos: additionall -> additional and externall -> external)

tapeinosyne commented:
I feel we should open another RFC for the sole introduction of the `changelog` key. That much is uncontroversional, indisputably useful, and there is a strong precedent in `readme`, whereas detailed changelog formats require far more discussion. If we can agree that any guidelines can be established incrementally as conventions emerge, instead of enforcing a standard from the start, there would be no risk of breaking changes; even then, stabilization could still be delayed.

newpavlov commented:
@dbrgn 
>I think that's overly strict and that Cargo should not add such a requirement. It will only result in workarounds, like not adding a heading for the unreleased changes.

I don't think it makes much sense to have "unreleased" section in the changelog for _published_ crate, especially if changelogs will be initially rendered as readme without any post-processing.

>Why is that? It only makes it harder for projects that have had a few releases before their first crates.io release to publish. I see no advantage in adding this restriction.

Hm, you are right, I haven't considered cases when crate can be initially developed without getting published on crates.io. I've removed this requirement.

@tapeinosyne 
This "detailed changelog format" currently essentially boils down to "headline must start with the version", plus a safeguard to help crate authors remember to update their changelogs before publishing new crate version, which can be disabled if needed. As I see it, what you propose is to have the second readme, which seriously hinders any attempts to develop tooling around changelogs. And no, I don't think such approach will result in emergence of conventions of any sort.

Also it's always useful to keep in mind that: "nothing is more permanent than the temporary".

dbrgn commented:
> I don't think it makes much sense to have "unreleased" section in the changelog for published crate, especially if changelogs will be initially rendered as readme without any post-processing.

Hm, that's a good point. Some workflows will need to change, as some people add the unreleased section at the same time they release the previous version. But that change can just as well be done *after* the release commit.

demurgos commented on text/0000-crate-changelogs.md:
```diff
@@ -0,0 +1,237 @@
+[package]
+name = "foo"
+version = "0.1.0"
+changelog = "CHNAGELOG.md"
```
There's a small typo here. `CHNAGELOG.md` -> `CHANGELOG.md`

newpavlov replied:
Thank you for noticing!

ehuss commented:
We discussed this RFC in the Cargo team meeting, and we feel like we probably don't want to move forward with this at this time. There are some points, such as the proposal for a single format that could be machine-readable that we think may be a little too strict. If the proposal then reduces to just a link that will be displayed, then it seems to not bring much value over placing a link within a README.  

We acknowledge that there is a desire to encourage crate authors to provide a changelog, and to make it easier to access that information. We thank everyone for their work and discussion. There may be some solution to improve communication of release notes, but it's not clear what that would look like.

@rfcbot fcp close

ishitatsuyuki commented:
@rfcbot reviewed

Eh2406 commented:
@rfcbot reviewed

smarnach commented:
@rfcbot reviewed

sgrif commented:
@rfcbot reviewed

On Sun, Sep 22, 2019, 1:56 PM Sven Marnach <notifications@github.com> wrote:

> @rfcbot <https://github.com/rfcbot> reviewed
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/2129?email_source=notifications&email_token=AALVMK367NYFUMMJNQ4GXLTQK7EV3A5CNFSM4DYOUGE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7JN3NI#issuecomment-533913013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALVMK6EUFCPYS3L72GLYVDQK7EV3ANCNFSM4DYOUGEQ>
> .
>

Turbo87 commented:
@rfcbot reviewed

While I do agree that crates with more than one release should probably have a changelog, I don't think enforcing them like that is something we should do. If we built some sort of rating system into crates.io that judges the quality of a crate we could consider it as part of the metric though.

jtgeibel commented:
@rfcbot reviewed

locks commented:
@rfcbot reviewed

georgefst commented:
Hmm, this is disappointing. The fact that so many crates don't provide any changelog is my number one frustration with the Rust ecosystem. As someone who only sporadically writes Rust, it tends to be that I come back to the language every few months, bump a load of dependencies, then spend ages trying to work out exactly what broke all of my code. I just need a migration guide!

I haven't read _all_ of the above (it's a long discussion!), but would it be fair to say that there was significant support here for the idea that Cargo should encourage crate authors to write human-readable changelogs, but that that got lost amid a lot of bikeshedding about requiring a more precise format? If so, is anyone willing to submit a minimal version of the proposal?

mathstuf commented:
> If so, is anyone willing to submit a minimal version of the proposal?

IMO, I think "use markdown", "use one of these blessed filenames (`CHANGELOG`, `CHANGELOG.md`, etc.)",  and "use the version number as a top-level heading (maybe with a `v` prefix)" is about as structured as anyone is going to agree on.

kxxt commented:
> If so, is anyone willing to submit a minimal version of the proposal?

I haven't read all the discussion. But I often find myself jumping through hops and hops just to figure out where the changelog is. (Editor -> crates.io ->  repo/website -> search for changelog). I am planning to submit a reduced version of this RFC.

I don't think we need to include a markdown changelog in the crate itself, as the included changelog will become outdated and the size of that markdown will grow over time. An external link should be enough.
